internal/ui/concurrency/process_download_concurrency_model.go

```diff
@@ -0,0 +1,274 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ui_concurrency
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/charmbracelet/bubbles/progress"
+	tea "github.com/charmbracelet/bubbletea"
+	"github.com/charmbracelet/lipgloss"
+
+	"tidbcloud-cli/internal/util"
+)
+
+var helpStyle = lipgloss.NewStyle().Foreground(lipgloss.Color("#626262")).Render
+
+const (
+	padding  = 2
+	maxWidth = 80
+)
+
+type ProgressMsg struct {
+	id      int
+	percent float64
+}
+
+func NewProgressMsg(id int, percent float64) ProgressMsg {
+	return ProgressMsg{id: id, percent: percent}
+}
+
+type ProgressErrMsg struct {
+	id  int
+	err error
+}
+
+func NewProgressErrMsg(id int, err error) ProgressErrMsg {
+	return ProgressErrMsg{id: id, err: err}
+}
+
+type FileJob struct {
+	id      int
+	path    string
+	name    string
+	url     string
+	process progress.Model
+	err     error
+}
+
+type JobInfo struct {
+	idToJob       map[int]*FileJob
+	viewJobs      []*FileJob
+	finishedCount int
+	total         int
+	lock          sync.Mutex
+}
+
+type Model struct {
+	concurrency int
+	jobsCh      chan *FileJob
+	jobInfo     *JobInfo
+	width       int
+	once        sync.Once
+	onProgress  func(int, float64)
+	onError     func(int, error)
+	Interrupted bool
+}
+
+type URLMsg struct {
+	Name string
+	Path string
+	Url  string
+}
+
+func NewModel(urls []URLMsg, onProgress func(int, float64), onError func(int, error), concurrency int) Model {
+	jobs := make(chan *FileJob, len(urls))
+	idToJob := make(map[int]*FileJob)
+
+	for i, url := range urls {
+		job := &FileJob{id: i, name: url.Name, url: url.Url, path: url.Path}
+		idToJob[i] = job
+	}
+
+	jobInfo := &JobInfo{
+		idToJob:       idToJob,
+		viewJobs:      make([]*FileJob, 0, len(urls)),
+		finishedCount: 0,
+		total:         len(urls),
+	}
+	return Model{
+		jobsCh:      jobs,
+		jobInfo:     jobInfo,
+		concurrency: concurrency,
+		onProgress:  onProgress,
+		onError:     onError,
+	}
+}
+
+func NewDefaultModel(urls []URLMsg, onProgress func(int, float64), onError func(int, error)) Model {
+	return NewModel(urls, onProgress, onError, 2)
+}
+
+func (m Model) Init() tea.Cmd {
+	return nil
+}
+
+func (m *Model) InitPool() {
```

Comment:
why here is a pointer receiver and others are value receiver?

Reply:
-  InitPool(): it use sync.Once defined in the model, it can not execute only once when using value receiver.
- others: we must use value receiver to implement the model interface.